### PR TITLE
Advantage scope

### PR DESCRIPTION
**Pull request by ModBoyex**
Please switch to **Preview** and select one of the following links:

* [develop](?template=develop.md)
* [main](?template=main.md)

Once switched to the correct template, you can save the link as a bookmark. Keep in mind that switching templates will remove all already entered data within this issue.

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -19,13 +28,15 @@
...
+  public static boolean isReplayMode = false;
+  public static double kDefaultPeriod = 0.02;
 
   //RoboRIO serial numbers
-  public static final String competitionRobotSerial = "03161743";
-  public static final String practiceRobotSerial = "03064df0";
-  public static final boolean isCompetitionRobot = !HALUtil.getSerialNumber().equals(practiceRobotSerial);
+  public static String competitionRobotSerial = "03161743";;
+  public static String practiceRobotSerial = "03064df0";
+  public static boolean isCompetitionRobot = !HALUtil.getSerialNumber().equals(practiceRobotSerial);
```
Comment (zbrotherton2158) [inline src/main/java/frc/robot/Robot.java:39]: Undo these changes

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -19,13 +28,15 @@
  * the package after creating this project, you must also update the build.gradle file in the
  * project.
  */
-public class Robot extends TimedRobot {
-  public static final boolean isCompetition = true;
+public class Robot extends LoggedRobot {
+  public static boolean isCompetition = true;
+  public static boolean isReplayMode = false;
+  public static double kDefaultPeriod = 0.02;
```
Comment (zbrotherton2158) [inline src/main/java/frc/robot/Robot.java:34]: These should all be final

File: src/main/java/frc/robot/Robot.java [hunk 3]
```diff
@@ -50,6 +61,23 @@ public class Robot extends TimedRobot {
...
   @Override
   public void robotInit() {
+    Logger.getInstance().recordMetadata("ProjectName", "MyProject"); // Set a metadata value
+
+    if (isReal()) {
+        Logger.getInstance().addDataReceiver(new WPILOGWriter("/media/sda1/")); // Log to a USB stick
+        Logger.getInstance().addDataReceiver(new NT4Publisher()); // Publish data to NetworkTables
+        new PowerDistribution(1, ModuleType.kRev); // Enables power distribution logging
+    } else {
+        setUseTiming(false); // Run as fast as possible
```
Comment (zbrotherton2158) [inline src/main/java/frc/robot/Robot.java:71]: Should be within isReplayMode()

**Review by zbrotherton2158 — APPROVED** [review]
Let's try to merge ASAP and make 3D a separate feature

File: build.gradle [hunk 4]
```diff
@@ -97,3 +101,27 @@ wpi.java.configureTestTasks(test)
...
+    exclude group: "edu.wpi.first.wpilibj"
+}
+
+project.compileJava.dependsOn(createVersionFile)
+gversion {
+  srcDir       = "src/main/java/"
+  classPackage = "frc.robot"
+  className    = "BuildConstants"
+  dateFormat   = "yyyy-MM-dd HH:mm:ss z"
+  timeZone     = "America/New_York" // Use preferred time zone
```
Comment (zbrotherton2158) [inline build.gradle:125]: Might want to change the time zone, I think typing CST would work? (Although our timezone shifts to CDT mid competition)

**Review by zbrotherton2158 — APPROVED** [review]

**Review by azvanderpas — APPROVED** [review]
